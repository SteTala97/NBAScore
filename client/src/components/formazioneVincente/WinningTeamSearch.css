/* STANDING DROP-DOWN MENU STYLE */

:root {
    /* --gray: #00db00; */
    --darkgray: #C9082A; /* Parte sinistra select box */
  }
  
  select {
    font-family: system-ui,-apple-system,system-ui,"Helvetica Neue",Helvetica,Arial,sans-serif;
    appearance: none;
    font-weight: 600;
    font-size: 16px;
    outline: 0;
    border: 0;
    box-shadow: none;
    flex: 1;
    padding: 0 1em;
    color: #fff;
    background-color: var(--darkgray);
    background-image: none;
    cursor: pointer;
  }
  
  select::-ms-expand {
    display: none;
  }
  /* Custom Select wrapper */
  .select {
    margin-right: 30px;
    position: relative;
    display: flex;
    width: 250px;
    height: 3em;
    border-radius: .25rem;
    overflow: hidden;
  }
  /* Arrow */
  .select::after {
    content: '\25BC';
    position: absolute;
    top: 0;
    right: 0;
    padding: 1em;
    background-color: #C9082A; /* Freccia a destra */
    transition: .25s all ease;
    pointer-events: none;
  }
  /* Transition */
  .select:hover::after {
    color: #ffffff;
  }

  option { 
    font-size:16px; 
  }
  
  .search-menu {
    margin-top: 138px;
    display: flex;
    justify-content: center;
    align-items: center;
  }
  
  
  .search-button {
  font-family: system-ui,-apple-system,system-ui,"Helvetica Neue",Helvetica,Arial,sans-serif;
  font-size: 16px;
  align-items: center;
  background-color: #C9082A;
  border: 1px solid transparent;
  border-radius: .25rem;
  box-shadow: rgba(0, 0, 0, 0.02) 0 1px 3px 0;
  box-sizing: border-box;
  color: #FFFFFF;
  cursor: pointer;
  display: inline-flex;
  font-weight: 600;
  justify-content: center;
  line-height: 1.25;
  min-height: 3rem;
  padding: calc(.875rem - 1px) calc(1.5rem - 1px);
  position: relative;
  text-decoration: none;
  transition: all 250ms;
  user-select: none;
  -webkit-user-select: none;
  touch-action: manipulation;
  vertical-align: baseline;
  width: auto;
  }