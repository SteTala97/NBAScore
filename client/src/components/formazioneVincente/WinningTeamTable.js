import React from 'react';

import Table from '@mui/material/Table';
import TableBody from '@mui/material/TableBody';
import TableCell from '@mui/material/TableCell';
import TableContainer from '@mui/material/TableContainer';
import TableHead from '@mui/material/TableHead';
import TableRow from '@mui/material/TableRow';
import List from '@mui/material/List';
import ListItem from '@mui/material/ListItem';
import ListItemButton from '@mui/material/ListItemButton';
import ListItemText from '@mui/material/ListItemText';

// Visualization of the winning team
function WinningTeamTable(props) {
    // Research not started
    if(!props.searchStarted)
        return (null)
    // Research started and error in the result of the first API call
    else if(props.winningGamesError)
        return <p className="no-data-message"> Error: {props.winningGamesError.message} </p>
    // Research started and error in the result of the second API call
    else if(props.statisticsError)
        return <p className="no-data-message"> Error: {props.statisticsError.message} </p>
    // Research started and no error, but one of the two APIs returns an empty response
    else if(props.winningGames.length === 0 || props.statistics.length === 0)
        return <p className="no-data-message"> Results not available for these parameters </p>
    // Research started and no error, but the API of the players statistics contains wrong data
    // In fact, the initial team for every game is composed by 5 players, so if we have a length that isn't a multiple of 5 it means we have, for example, 6 or 4 players per team
    else if(props.statistics.length % 5 !== 0)
        return <p className="no-data-message"> Results not available for these parameters, because the remote API contains wrong data </p>
    // Research started, no error and length of statistics is fine
    else {
        // The players statistics API returns a list where every element contains the data for a certain player in a certain game
        // Consecutive elements refer to different players in the same game
        // Since in the server we kept only players with a position, it means that we have to have exactly 5 consecutive elements referring to the same game
        // If it's not, it means the API of the players statistics contains wrong data
        var checkGamesId = true;
        for(var i = 0; i < props.statistics.length; i = i + 5) {
            if(props.statistics[i].gameId !== props.statistics[i + 1].gameId ||
                props.statistics[i + 1].gameId !== props.statistics[i + 2].gameId ||
                props.statistics[i + 2].gameId !== props.statistics[i + 3].gameId ||
                props.statistics[i + 3].gameId !== props.statistics[i + 4].gameId)
                    checkGamesId = false;

            if(!checkGamesId)
                return <p className="no-data-message"> Results not available for these parameters, because the remote API contains wrong data </p>
        }

        // Dictionary where the key is the concatenation of the IDs of the players of the team and the value is the number of wins of that team
        var dictionary = {};

        for(i = 0; i < props.statistics.length; i = i + 5) {
            // Considering only the games where the team has won
            if(props.winningGames.includes(props.statistics[i].gameId)) {
                var key = props.statistics[i].playerId + " " +
                          props.statistics[i + 1].playerId + " " +
                          props.statistics[i + 2].playerId + " " +
                          props.statistics[i + 3].playerId + " " +
                          props.statistics[i + 4].playerId;

                var array = key.split(' ');
                var array_sorted = array.sort();
                key = array_sorted.join(' ');

                // Already in the dictionary, so I increment the value by one
                if(key in dictionary)
                    dictionary[key] = dictionary[key] + 1;
                // First time, so I set the value to one
                else
                    dictionary[key] = 1;
            }
        }

        var max_win = 0;
        var total_win = 0;
        var team_max_win = "";
        
        // Selecting the maximum number of wins (and which formation did it) and the total number of wins of the team
        for(key in dictionary) {
            total_win = total_win + dictionary[key];
            if(dictionary[key] > max_win) {
                max_win = dictionary[key];
                team_max_win = key;
            }
        }
        team_max_win = team_max_win.split(' ');

        var team_max_win_names = [];
        var toSearch = true;

        // Searching the name of the players of the winning team
        for(i = 0; i < team_max_win.length; i++) {
            toSearch = true;
            for(var j = 0; j < props.statistics.length && toSearch; j++) {
                if(team_max_win[i] == props.statistics[j].playerId) {
                    team_max_win_names.push(props.statistics[j].player);
                    toSearch = false;
                }
            }
        }

        return (
            <TableContainer align="center">
                <Table id="winningTeamTable" aria-label="simple table" cellSpacing="0" sx={{ height: "max-content", minWidth: 100, maxWidth: 300, marginTop: 6}}>
                    <TableHead>
                        <TableRow key="1">
                            <TableCell align="center"><strong>This is the most winning formation: it has won {max_win} games on a total of {total_win} games won by this team</strong></TableCell>
                        </TableRow>
                    </TableHead>
                    <TableBody>                        
                        <TableRow>
                            <TableCell>
                                <List>
                                    <ListItem disablePadding>
                                        <ListItemButton onClick={() => { props.setPlayerInfo(true); props.setPlayerId(team_max_win[0]) }}>
                                            <ListItemText align="center" primary={team_max_win_names[0]} />
                                        </ListItemButton>
                                    </ListItem>
                                    <ListItem disablePadding>
                                        <ListItemButton onClick={() => { props.setPlayerInfo(true); props.setPlayerId(team_max_win[1]) }}>
                                            <ListItemText align="center" primary={team_max_win_names[1]} />
                                        </ListItemButton>
                                    </ListItem>
                                    <ListItem disablePadding>
                                        <ListItemButton onClick={() => { props.setPlayerInfo(true); props.setPlayerId(team_max_win[2]) }}>
                                            <ListItemText align="center" primary={team_max_win_names[2]} />
                                        </ListItemButton>
                                    </ListItem>
                                    <ListItem disablePadding>
                                        <ListItemButton onClick={() => { props.setPlayerInfo(true); props.setPlayerId(team_max_win[3]) }}>
                                            <ListItemText align="center" primary={team_max_win_names[3]} />
                                        </ListItemButton>
                                    </ListItem>
                                    <ListItem disablePadding>
                                        <ListItemButton onClick={() => { props.setPlayerInfo(true); props.setPlayerId(team_max_win[4]) }}>
                                            <ListItemText align="center" primary={team_max_win_names[4]} />
                                        </ListItemButton>
                                    </ListItem>
                                </List>
                            </TableCell>            
                        </TableRow>
                    </TableBody>                    
                </Table>
            </TableContainer>
        );
    }
}

export default WinningTeamTable;
