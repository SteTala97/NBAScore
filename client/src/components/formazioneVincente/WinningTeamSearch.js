import React, { useEffect } from 'react';
import './WinningTeamSearch.css'
const axios = require('axios');

// Selection of the team and the season for the research
const WinningTeamSearch = (props) => {
    
    // When the component is loaded we require the list of the teams and the list of the seasons
    useEffect( () => {
        async function loadTeams() {
            await axios.get("https://nbascore.herokuapp.com/teams") 
                .then(function (response) {
                    props.setTeams(response.data);
                    props.setTeamError(null);
                })
                .catch(function (error) {
                    props.setTeamError(error);
                })
        }

        loadTeams();

        async function loadSeasons() {
            await axios.get("https://nbascore.herokuapp.com/seasons") 
                .then(function (response) {
                    props.setSeasons(response.data);
                    props.setSeasonError(null);
                })
                .catch(function (error) {
                    props.setSeasonError(error);
                })
        }

        loadSeasons();
    }, []);
        

    // Search the winning team
    async function search() {             
        // Missing input parameters
        if(props.team === "") {
            alert("You have to select a Team!");
        }
        else if(props.season === 0) {
            alert("You have to select a Season!");
        }
        // Call to the APIs needed to complete the research
        else{

            await axios.get("https://nbascore.herokuapp.com/gamesPerTeamAndSeason", { params: { season: props.season, team: props.team } })
                .then(function (response) {
                    props.setWinningGames(response.data);
                    props.setWinningGamesError(null);
                })
                .catch(function (error) {
                    props.setWinningGamesError(error);
                })

            await axios.get("https://nbascore.herokuapp.com/playersStatisticsPerTeamAndSeason", { params: { season: props.season, team: props.team } })
            .then(function (response) {
                props.setStatistics(response.data);
                props.setStatisticsError(null);
            })
            .catch(function (error) {
                props.setStatisticsError(error);
            })

            props.setSearchStarted(true);
        }
    }

    // Request to obtain the list of teams is failed
    if(props.teamError){
        return (
            <div>
                Service temporarily unavailable. Unable to connect to the remote API. 
                <br />
                {props.teamError.message}
            </div>
        );
    }
    // Request to obtain the list of seasons is failed
    else if(props.seasonError){
        return (
            <div>
                Service temporarily unavailable. Unable to connect to the remote API. 
                <br />
                {props.seasonError.message}
            </div>
        );
    }
    // Everything is ok, we can render the research component
    else {
        return (
            <div className='search-menu'>
                    <form className='select'>
                        <select data-testid="teamToSearch" id="teamToSearch" onChange={e=>props.setTeam(e.target.value)}>
                            <option value="" selected disabled>Select Team</option>
                            {
                                props.teams.map(team => (
                                    <option value={team.id}>{team.name}</option>
                                )) 
                            }
                        </select>
                    </form>
                    <form className='select'>
                        <select data-testid="seasonToSearch" id="seasonToSearch" onChange={e=>props.setSeason(e.target.value)}>
                            <option value="" selected disabled>Select Season</option>
                            {
                                props.seasons.map(season => (
                                    <option value={season}>{season}</option>
                                )) 
                            }
                        </select>
                    </form>
            <button className="search-button" id="search-button" onClick={search}>Search</button> 
            </div>
        )
    };    
}

export default WinningTeamSearch;