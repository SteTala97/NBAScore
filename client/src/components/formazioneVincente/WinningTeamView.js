import React, { useState } from 'react';
import WinningTeamSearch from './WinningTeamSearch.js';
import WinningTeamTable from './WinningTeamTable.js';
import PlayerInfo from '../visualizzazioneInformazioniGiocatore/PlayerInfo';

// Root page for the visualization of the winning team
function WinningTeamView () {
    const [team, setTeam] = useState("");
    const [teams, setTeams] = useState([]);
    const [teamError, setTeamError] = useState(null);

    const [season, setSeason] = useState(0);
    const [seasons, setSeasons] = useState([]);
    const [seasonError, setSeasonError] = useState(null);

    const [searchStarted, setSearchStarted] = useState(false);

    const [winningGames, setWinningGames] = useState([]);
    const [winningGamesError, setWinningGamesError] = useState(null);

    const [statistics, setStatistics] = useState([]);
    const [statisticsError, setStatisticsError] = useState(null);

    const [playerInfo, setPlayerInfo] = useState(false);
    const [playerId, setPlayerId] = useState(0);

    return (
        playerInfo === true ?
            <PlayerInfo playerId = {playerId} />
        :
        <React.Fragment>
            <WinningTeamSearch
                team = {team}
                setTeam = {setTeam}
                teams = {teams}
                setTeams = {setTeams}
                teamError = {teamError}
                setTeamError = {setTeamError}

                season = {season}
                setSeason = {setSeason}
                seasons = {seasons}
                setSeasons = {setSeasons}
                seasonError = {seasonError}
                setSeasonError = {setSeasonError}

                searchStarted = {searchStarted}
                setSearchStarted = {setSearchStarted}

                winningGames = {winningGames}
                setWinningGames = {setWinningGames}
                winningGamesError = {winningGamesError}
                setWinningGamesError = {setWinningGamesError}

                statistics = {statistics}
                setStatistics = {setStatistics}
                statisticsError = {statisticsError}
                setStatisticsError = {setStatisticsError}

                playerInfo = {playerInfo}
                setPlayerInfo = {setPlayerInfo}
                playerId = {playerId}
                setPlayerId = {setPlayerId}

             />
             <WinningTeamTable
                team = {team}
                setTeam = {setTeam}
                teams = {teams}
                setTeams = {setTeams}
                teamError = {teamError}
                setTeamError = {setTeamError}

                season = {season}
                setSeason = {setSeason}
                seasons = {seasons}
                setSeasons = {setSeasons}
                seasonError = {seasonError}
                setSeasonError = {setSeasonError}

                searchStarted = {searchStarted}
                setSearchStarted = {setSearchStarted}

                winningGames = {winningGames}
                setWinningGames = {setWinningGames}
                winningGamesError = {winningGamesError}
                setWinningGamesError = {setWinningGamesError}

                statistics = {statistics}
                setStatistics = {setStatistics}
                statisticsError = {statisticsError}
                setStatisticsError = {setStatisticsError}

                playerInfo = {playerInfo}
                setPlayerInfo = {setPlayerInfo}
                playerId = {playerId}
                setPlayerId = {setPlayerId}
             />
        </React.Fragment>
    );
}

export default WinningTeamView;