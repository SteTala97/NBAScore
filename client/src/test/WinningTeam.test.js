import React from 'react';

import { shallowToJson } from 'enzyme-to-json';
import { shallow } from 'enzyme';
import { mount } from 'enzyme';

import WinningTeamView from '../components/formazioneVincente/WinningTeamView';
import WinningTeamSearch from '../components/formazioneVincente/WinningTeamSearch';
import WinningTeamTable from '../components/formazioneVincente/WinningTeamTable';

describe("Test rendering componenti WinningTeam", () => {

    let initialProps;

    beforeEach( () => {
        initialProps = {
            team: "",
            teams: [],
            teamError: null,
            season : 0,
            seasons: [], 
            seasonError: null,
            searchStarted: false,
            winningGames: [],
            winningGamesError: null,
            statistics: [],
            statisticsError: null,
            playerInfo: false,
            playerId: 0,
            setTeam: jest.fn(x => initialProps.team = x),
            setTeams: jest.fn(x => initialProps.teams = x),
            setTeamError: jest.fn(x => initialProps.teamError = x),
            setSeason: jest.fn(x => initialProps.season = x),
            setSeasons: jest.fn(x => initialProps.seasons = x),
            setSeasonError: jest.fn(x => initialProps.seasonError = x),
            setSearchStarted: jest.fn(x => initialProps.searchStarted = x),
            setWinningGames: jest.fn(x => initialProps.winningGames = x),
            setWinningGamesError: jest.fn(x => initialProps.winningGamesError = x),
            setStatistics: jest.fn(x => initialProps.statistics = x),
            setStatisticsError: jest.fn(x => initialProps.statisticsError = x),
            setPlayerInfo: jest.fn(x => initialProps.playerInfo = x),
            setPlayerId: jest.fn(x => initialProps.playerId = x),
        };
    });
  
    it("Rendering del componente WinningTeamView senza crashare", () => {
        shallow(<WinningTeamView{...initialProps}/>);
    });

    it("Rendering del componente WinningTeamSearch senza crashare", () => {
        shallow(<WinningTeamSearch{...initialProps}/>);
    });

    it("Rendering del componente WinningTeamTable senza crashare", () => {
        shallow(<WinningTeamTable{...initialProps}/>);
    });

    it('Rendering dei menÃ¹ a tendina delle stagioni e delle squadre', () => {
        initialProps.seasons = [2019, 2020, 2021];
        initialProps.teams = ["Chicago Bulls", "Dallas Mavericks", "Phoenix Suns"];
    
        let wrapper = mount(<WinningTeamSearch{...initialProps}/>);
        let seasonOptions = wrapper.find('#seasonToSearch').find('option');
        let teamOptions = wrapper.find('#teamToSearch').find('option');
    
        expect(seasonOptions.length).toEqual(initialProps.seasons.length + 1);
    
        let i = 0;
        seasonOptions.forEach((node) => {
          const texts = node.text();
          if(i == 0) {
            expect(texts).toEqual('Select Season'); 
          }
          else {
            expect(texts).toEqual(initialProps.seasons[i-1].toString()); 
          }
          i++;
        });
    
        expect(teamOptions.length).toEqual(initialProps.teams.length + 1);

        i = 0;
        teamOptions.forEach((node) => {
          const texts = node.text();
          if(i == 0) {
            expect(texts).toEqual('Select Team'); 
          }
          else {
            expect(texts).toEqual("");
          }
          i++;
        });   
      });

      it("Rendering del bottone Ricerca senza crashare con controllo label", () => {
        const wrapper = mount(<WinningTeamSearch{...initialProps}/>);
        const label = wrapper.find("#search-button").text();
        expect(label).toEqual("Search");
      });

      it("Snapshot componente WinningTeamView", () => {
        const tree = shallow(<WinningTeamView{...initialProps}/>);
        expect(shallowToJson(tree)).toMatchSnapshot();
      });

      it("Snapshot componente WinningTeamSearch", () => {
        const tree = shallow(<WinningTeamSearch{...initialProps}/>);
        expect(shallowToJson(tree)).toMatchSnapshot();
      });

      it("Snapshot componente WinningTeamTable", () => {
        const tree = shallow(<WinningTeamTable{...initialProps}/>);
        expect(shallowToJson(tree)).toMatchSnapshot();
      });
});