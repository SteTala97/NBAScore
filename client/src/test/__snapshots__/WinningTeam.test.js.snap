// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test rendering componenti WinningTeam Snapshot componente WinningTeamSearch 1`] = `
<div
  className="search-menu"
>
  <form
    className="select"
  >
    <select
      data-testid="teamToSearch"
      id="teamToSearch"
      onChange={[Function]}
    >
      <option
        disabled={true}
        selected={true}
        value=""
      >
        Select Team
      </option>
    </select>
  </form>
  <form
    className="select"
  >
    <select
      data-testid="seasonToSearch"
      id="seasonToSearch"
      onChange={[Function]}
    >
      <option
        disabled={true}
        selected={true}
        value=""
      >
        Select Season
      </option>
    </select>
  </form>
  <button
    className="search-button"
    id="search-button"
    onClick={[Function]}
  >
    Search
  </button>
</div>
`;

exports[`Test rendering componenti WinningTeam Snapshot componente WinningTeamTable 1`] = `""`;

exports[`Test rendering componenti WinningTeam Snapshot componente WinningTeamView 1`] = `
<Fragment>
  <WinningTeamSearch
    playerId={0}
    playerInfo={false}
    searchStarted={false}
    season={0}
    seasonError={null}
    seasons={Array []}
    setPlayerId={[Function]}
    setPlayerInfo={[Function]}
    setSearchStarted={[Function]}
    setSeason={[Function]}
    setSeasonError={[Function]}
    setSeasons={[Function]}
    setStatistics={[Function]}
    setStatisticsError={[Function]}
    setTeam={[Function]}
    setTeamError={[Function]}
    setTeams={[Function]}
    setWinningGames={[Function]}
    setWinningGamesError={[Function]}
    statistics={Array []}
    statisticsError={null}
    team=""
    teamError={null}
    teams={Array []}
    winningGames={Array []}
    winningGamesError={null}
  />
  <WinningTeamTable
    playerId={0}
    playerInfo={false}
    searchStarted={false}
    season={0}
    seasonError={null}
    seasons={Array []}
    setPlayerId={[Function]}
    setPlayerInfo={[Function]}
    setSearchStarted={[Function]}
    setSeason={[Function]}
    setSeasonError={[Function]}
    setSeasons={[Function]}
    setStatistics={[Function]}
    setStatisticsError={[Function]}
    setTeam={[Function]}
    setTeamError={[Function]}
    setTeams={[Function]}
    setWinningGames={[Function]}
    setWinningGamesError={[Function]}
    statistics={Array []}
    statisticsError={null}
    team=""
    teamError={null}
    teams={Array []}
    winningGames={Array []}
    winningGamesError={null}
  />
</Fragment>
`;
